ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LSM.hs

```diff
@@ -0,0 +1,478 @@
+  Session m ->
+  l EmptyMK ->
+  LedgerTables l ValuesMK ->
+  m (ResourceKey m, UTxOTable m)
+tableFromValuesMK rr session st (LedgerTables (ValuesMK values)) = do
+  res@(_, table) <-
+    allocate
+      rr
+      ( \_ ->
+          LSM.newTableWith (LSM.defaultTableConfig{LSM.confFencePointerIndex = LSM.OrdinaryIndex}) session
+      )
+      LSM.closeTable
+  let sz = min 1000 (Map.size values)
+  vec <- VM.unsafeNew sz -- can use unsafe
+  (n, vec') <- ifoldlM (go table) (0, vec) values
+  case n of
+    0 -> pure ()
+    _ -> do
+      vec'' <- V.unsafeFreeze vec'
+      LSM.inserts table vec''
+  pure res
+ where
+  go table k (n, vec) v
+    | n == 1000 =
+        do
+          vec' <- V.unsafeFreeze vec -- can use unsafe because vec won't be used afterwards
+          LSM.inserts table vec'
+          vec'' <- VM.unsafeNew 1000
+          pure (0, vec'')
+    | otherwise =
+        do
+          VM.unsafeWrite vec n (toTxInBytes (Proxy @l) k, toTxOutBytes st v, Nothing) -- can use unsafe
+          pure (n + 1, vec)
+
+snapshotManager ::
+  ( IOLike m
+  , LedgerDbSerialiseConstraints blk
+  , LedgerSupportsProtocol blk
+  ) =>
+  Session m ->
+  Complete LedgerDbArgs m blk ->
+  SnapshotManager m m blk (StateRef m (ExtLedgerState blk))
+snapshotManager session args =
+  snapshotManager'
+    session
+    (configCodec . getExtLedgerCfg . ledgerDbCfg $ lgrConfig args)
+    (LedgerDBSnapshotEvent >$< lgrTracer args)
+    (lgrHasFS args)
+
+snapshotManager' ::
+  ( IOLike m
+  , LedgerDbSerialiseConstraints blk
+  , LedgerSupportsProtocol blk
+  ) =>
+  Session m ->
+  CodecConfig blk ->
+  Tracer m (TraceSnapshotEvent blk) ->
+  SomeHasFS m ->
+  SnapshotManager m m blk (StateRef m (ExtLedgerState blk))
+snapshotManager' session ccfg tracer fs =
+  SnapshotManager
+    { listSnapshots = defaultListSnapshots fs
+    , deleteSnapshot = implDeleteSnapshot session fs tracer
+    , takeSnapshot = implTakeSnapshot ccfg tracer fs
+    }
+
+newLSMLedgerTablesHandle ::
+  forall m l.
+  ( IOLike m
+  , HasLedgerTables l
+  , IndexedMemPack (l EmptyMK) (TxOut l)
+  , ShowProxy (LedgerBlock l)
+  ) =>
+  Tracer m V2.FlavorImplSpecificTrace ->
+  ResourceRegistry m ->
+  (ResourceKey m, UTxOTable m) ->
+  m (LedgerTablesHandle m l)
+newLSMLedgerTablesHandle tracer rr (resKey, t) = do
+  traceWith tracer V2.TraceLedgerTablesHandleCreate
+  pure
+    LedgerTablesHandle
+      { close = do
+          Monad.void $ release resKey
+          traceWith tracer V2.TraceLedgerTablesHandleClose
+      , duplicate = do
+          table <- allocate rr (\_ -> LSM.duplicate t) LSM.closeTable
+          newLSMLedgerTablesHandle tracer rr table
+      , read = \st (LedgerTables (KeysMK keys)) -> do
+          vec <- VM.unsafeNew (Set.size keys)
+          _ <-
+            Foldable.foldl'
+              (\m x -> m >>= \i -> VM.write vec i (toTxInBytes (Proxy @l) x) >> pure (i + 1))
+              (pure 0)
+              keys
+          vec' <- V.freeze vec
```

Comment:
There is [`create`](https://hackage-content.haskell.org/package/vector-0.13.2.0/docs/Data-Vector-Generic.html#v:create) to do this safely (and even more efficiently, as it can use `unsafeFreeze`).

Also, are we sure this is faster than
```haskell
V.fromListN (Set.size) $ toTxInBytes (Proxy @l) <$> Set.toList keys 
```
to warrant the extra complexity?

Reply:
Yeah, it is depressing that one has to make these kinds of judgement calls of simplicity vs (potential) predictable performance. *When* convenient staging for predictable code generation :stuck_out_tongue:

(Using `create` should be a clear win though.)

---

Also, looking at this again, using `foldM_` seems slightly cleaner